ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java

```diff
@@ -668,4 +694,64 @@ public JoinKey rewriteWith(
       return this;
     }
   }
+
+  private static final class ForeignJoinKey implements JoinKey {
+    private final ColumnName keyColumn;
+    private final ColumnName foreignKeyColumn;
+    private final ImmutableList<? extends ColumnReferenceExp> leftSourceKeyColumns;
+
+    static JoinKey of(final ColumnName keyColumn,
+                      final ColumnName foreignKeyColumn,
+                      final Collection<QualifiedColumnReferenceExp> leftSourceKeyColumns) {
+      return new ForeignJoinKey(keyColumn, foreignKeyColumn, leftSourceKeyColumns);
+    }
+
+    private ForeignJoinKey(final ColumnName keyColumn,
+                           final ColumnName foreignKeyColumn,
+                           final Collection<? extends ColumnReferenceExp> viableKeyColumns) {
+      this.keyColumn = requireNonNull(keyColumn, "keyColumn");
+      this.foreignKeyColumn = requireNonNull(foreignKeyColumn, "foreignKeyColumn");
+      this.leftSourceKeyColumns = ImmutableList
+          .copyOf(requireNonNull(viableKeyColumns, "viableKeyColumns"));
+    }
+
+    @Override
+    public boolean isSynthetic() {
+      return false;
+    }
+
+    @Override
+    public boolean isForeignKey() {
+      return true;
+    }
+
+    @Override
+    public List<? extends Expression> getOriginalViableKeys(final LogicalSchema schema) {
+      return leftSourceKeyColumns;
+    }
+
+    @Override
+    public List<? extends Expression> getAllViableKeys(final LogicalSchema schema) {
```

Comment:
As discussed offline, I still don't think we should be calling this method (or the one above) from this class at all. I would prefer to throw an error if someone tries to do this, but for the sake of getting this feature in, I'm OK with merging this if it works and we have plans to clean it up later.